libcudacxx/include/cuda/__iterator/transform_output_iterator.h

```diff
@@ -0,0 +1,500 @@
+        _CUDA_VSTD::is_assignable_v<_CUDA_VSTD::iter_reference_t<_Iter>, _CUDA_VSTD::invoke_result_t<_Fn&, _Arg>>)
+  _CCCL_API constexpr __transform_output_proxy& operator=(_Arg&& __arg) noexcept(
+    _CUDA_VSTD::is_nothrow_invocable_v<_Fn&, _Arg>
+    && _CUDA_VSTD::is_nothrow_assignable_v<_CUDA_VSTD::iter_reference_t<_Iter>, _CUDA_VSTD::invoke_result_t<_Fn&, _Arg>>)
+  {
+    *__iter_ = _CUDA_VSTD::invoke(__func_, _CUDA_VSTD::forward<_Arg>(__arg));
+    return *this;
+  }
+
+  _CCCL_EXEC_CHECK_DISABLE
+  _CCCL_TEMPLATE(class _Arg)
+  _CCCL_REQUIRES((!_CUDA_VSTD::is_same_v<_CUDA_VSTD::remove_cvref_t<_Arg>, __transform_output_proxy>)
+                   _CCCL_AND _CUDA_VSTD::is_invocable_v<const _Fn&, _Arg> _CCCL_AND
+                     _CUDA_VSTD::is_assignable_v<_CUDA_VSTD::iter_reference_t<const _Iter>,
+                                                 _CUDA_VSTD::invoke_result_t<const _Fn&, _Arg>>)
+  _CCCL_API constexpr const __transform_output_proxy& operator=(_Arg&& __arg) const
+    noexcept(_CUDA_VSTD::is_nothrow_invocable_v<const _Fn&, _Arg>
+             && _CUDA_VSTD::is_nothrow_assignable_v<_CUDA_VSTD::iter_reference_t<const _Iter>,
+                                                    _CUDA_VSTD::invoke_result_t<const _Fn&, _Arg>>)
+  {
+    *__iter_ = _CUDA_VSTD::invoke(__func_, _CUDA_VSTD::forward<_Arg>(__arg));
+    return *this;
+  }
+};
+
+//! \addtogroup iterators
+//! \{
+
+//!! \addtogroup fancyiterator Fancy Iterators
+//!  \ingroup iterators
+//!  \{
+
+//! @brief \p transform_output_iterator is an iterator which represents a pointer into a range of values after
+//! transformation by a function. This iterator is useful for creating a range filled with the result of applying an
+//! operation to another range without either explicitly storing it in memory, or explicitly executing the
+//! transformation. Using
+//! \p transform_output_iterator facilitates kernel fusion by deferring the execution of a transformation until the
+//! value is needed while saving both memory capacity and bandwidth.
+//!
+//! The following code snippet demonstrates how to create a \p transform_output_iterator which represents the result of
+//! \c sqrtf applied to the contents of a \p thrust::device_vector.
+//!
+//! @code
+//! #include <cuda/iterator>
+//! #include <thrust/device_vector.h>
+//!
+//! struct square_root
+//! {
+//!   __host__ __device__
+//!   float operator()(float x) const
+//!   {
+//!     return sqrtf(x);
+//!   }
+//! };
+//!
+//! int main()
+//! {
+//!   thrust::device_vector<float> v{1.0f, 4.0f, 9.0f, 16.0f};
+//!
+//!   using FloatIterator = thrust::device_vector<float>::iterator;
+//!
+//!   cuda::transform_output_iterator iter(v.begin(), square_root{});
+//!
+//!   *iter;   // returns 1.0f
+//!   iter[0]; // returns 1.0f;
+//!   iter[1]; // returns 2.0f;
+//!   iter[2]; // returns 3.0f;
+//!   iter[3]; // returns 4.0f;
+//!
+//!   // iter[4] is an out-of-bounds error
+//! }
```

Comment:
Important: this example does not work: https://godbolt.org/z/YPW1x6qrh

`thrust::transform_output_iterator` has a different example. Please use this one:
https://github.com/NVIDIA/cccl/blob/584fedba37077b0040cccf532229daa2d80ba312/thrust/thrust/iterator/transform_output_iterator.h#L115-L133